src/compiler/factory.ts

```diff
@@ -3886,6 +3886,27 @@ namespace ts {
     }
 
     /**
+     *  [Per the spec](https://tc39.github.io/ecma262/#prod-ExportDeclaration), `export default` accepts _AssigmentExpression_ but
+     *  has a lookahead restriction for `function`, `async function`, and `class`.
+     *
+     * Basically, that means we need to parenthesize in the following cases:
+     *
+     * - BinaryExpression of CommaToken
+     * - CommaList (synthetic list of multiple comma expressions)
+     * - FunctionExpression
+     * - ClassExpression
+     */
+    export function parenthesizeDefaultExpression(e: Expression) {
+        const check = skipPartiallyEmittedExpressions(e);
+        return (check.kind === SyntaxKind.ClassExpression ||
+            check.kind === SyntaxKind.FunctionExpression ||

```

Comment:
Is parenthesizing `FunctionExpression` and `ClassExpression` enough to stop them from being seen as `HoistableDeclaration`/`ClassDeclaration`? 🤔

Reply:
Yes. Once you add parens, you are no longer in a declaration.

Reply:
Should be. Anything that doesn't work that way is parsing incorrectly.